javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -37,83 +39,101 @@
 
     private static final long serialVersionUID = 1L;
 
-    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(Arrays.empty(), 0, 0, 0);
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(obj(), obj().empty(), 0, 0, 0);
     @SuppressWarnings("unchecked")
     static <T> BitMappedTrie<T> empty() { return (BitMappedTrie<T>) EMPTY; }
 
-    private final Object[] array;
+    final ArrayType<T> arrayType;
+    private final Object array;
     private final int offset, length;
     private final int depthShift;
 
-    private BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+    private BitMappedTrie(ArrayType<T> arrayType, Object array, int offset, int length, int depthShift) {
+        this.arrayType = arrayType;
         this.array = array;
         this.offset = offset;
         this.length = length;
         this.depthShift = depthShift;
 
         assert length <= treeSize(BRANCHING_FACTOR, depthShift);
-        assert (EMPTY == null) || ((length > 0) && (array.length > 0));
+        assert (EMPTY == null) || ((length > 0) && (ArrayType.of(array).lengthOf(array) > 0));
     }
 
     private static int treeSize(int branchCount, int depthShift) {
         final int fullBranchSize = 1 << depthShift;
         return branchCount * fullBranchSize;
     }
 
-    static <T> BitMappedTrie<T> ofAll(Object[] array) {
-        final int size = array.length;
+    static <T> BitMappedTrie<T> ofAll(Object array) {
+        final ArrayType<T> type = ArrayType.of(array);
+        final int size = type.lengthOf(array);
         if (size == 0) {
             return empty();
         } else {
             int shift = 0;
-            for (; array.length > BRANCHING_FACTOR; shift += BRANCHING_BASE) {
-                array = grouped(array, BRANCHING_FACTOR);
+            for (ArrayType<T> t = type; t.lengthOf(array) > BRANCHING_FACTOR; shift += BRANCHING_BASE) {
+                array = t.grouped(array, BRANCHING_FACTOR);
+                t = obj();
```

Comment:
yeah, trying to achieve a good signal-to-noise ratio in a language that doesn't want me to :)